File: Stripe/STPPaymentMethodBillingDetails.m
//
//  STPPaymentMethodBillingDetails.m
//  Stripe
//
//  Created by Yuki Tokuhiro on 3/5/19.
//  Copyright © 2019 Stripe, Inc. All rights reserved.
//

#import "STPPaymentMethodBillingDetails.h"

#import "STPPaymentMethodAddress.h"
#import "NSDictionary+Stripe.h"

@interface STPPaymentMethodBillingDetails ()

@property (nonatomic, copy, nonnull, readwrite) NSDictionary *allResponseFields;

@end

@implementation STPPaymentMethodBillingDetails

- (NSString *)description {
    NSArray *props = @[
                        // Object
                        [NSString stringWithFormat:@"%@: %p", NSStringFromClass([self class]), self],
                        
                        // Properties
                        [NSString stringWithFormat:@"name = %@", self.name],
                        [NSString stringWithFormat:@"phone = %@", self.phone],
                        [NSString stringWithFormat:@"email = %@", self.email],
                        [NSString stringWithFormat:@"address = %@", self.address],
                        ];
    return [NSString stringWithFormat:@"<%@>", [props componentsJoinedByString:@"; "]];
}

#pragma mark - STPFormEncodable
@synthesize additionalAPIParameters;

+ (nonnull NSDictionary *)propertyNamesToFormFieldNamesMapping {
    return @{
             NSStringFromSelector(@selector(address)): @"address",
             NSStringFromSelector(@selector(email)): @"email",
             NSStringFromSelector(@selector(name)): @"name",
             NSStringFromSelector(@selector(phone)): @"phone",
             };
}

+ (nullable NSString *)rootObjectName {
    return nil;
}

#pragma mark - STPAPIResponseDecodable

+ (nullable instancetype)decodedObjectFromAPIResponse:(nullable NSDictionary *)response {
    NSDictionary *dict = [response stp_dictionaryByRemovingNulls];
    if (!dict) {
        return nil;
    }
    STPPaymentMethodBillingDetails *billingDetails = [self new];
    billingDetails.allResponseFields = dict;
    billingDetails.address = [STPPaymentMethodAddress decodedObjectFromAPIResponse:[response stp_dictionaryForKey:@"address"]];
    billingDetails.email = [dict stp_stringForKey:@"email"];
    billingDetails.name = [dict stp_stringForKey:@"name"];
    billingDetails.phone = [dict stp_stringForKey:@"phone"];
    return billingDetails;
}

@end

File: Stripe/STPPaymentMethodBillingDetails.swift
//
//  STPPaymentMethodBillingDetails.swift
//  Stripe
//
//  Created by Yuki Tokuhiro on 3/5/19.
//  Copyright © 2019 Stripe, Inc. All rights reserved.
//

import Foundation

/// Billing information associated with a `STPPaymentMethod` that may be used or required by particular types of payment methods.
/// - seealso: https://stripe.com/docs/api/payment_methods/object#payment_method_object-billing_details
public class STPPaymentMethodBillingDetails: NSObject, STPAPIResponseDecodable, STPFormEncodable {
  @objc public var additionalAPIParameters: [AnyHashable: Any] = [:]

  /// Billing address.
  @objc public var address: STPPaymentMethodAddress?
  /// Email address.
  @objc public var email: String?
  /// Full name.
  @objc public var name: String?
  /// Billing phone number (including extension).
  @objc public var phone: String?
  @objc public private(set) var allResponseFields: [AnyHashable: Any] = [:]

  /// :nodoc:
  @objc public override var description: String {
    let props = [
      // Object
      String(format: "%@: %p", NSStringFromClass(STPPaymentMethodBillingDetails.self), self),
      // Properties
      "name = \(name ?? "")",
      "phone = \(phone ?? "")",
      "email = \(email ?? "")",
      "address = \(String(describing: address))",
    ]
    return "<\(props.joined(separator: "; "))>"
  }

  /// :nodoc:
  @objc public override required init() {
    super.init()
  }

  // MARK: - STPFormEncodable

  @objc
  public class func propertyNamesToFormFieldNamesMapping() -> [String: String] {
    return [
      NSStringFromSelector(#selector(getter:address)): "address",
      NSStringFromSelector(#selector(getter:email)): "email",
      NSStringFromSelector(#selector(getter:name)): "name",
      NSStringFromSelector(#selector(getter:phone)): "phone",
    ]
  }

  @objc
  public class func rootObjectName() -> String? {
    return nil
  }

  // MARK: - STPAPIResponseDecodable
  @objc
  public class func decodedObject(fromAPIResponse response: [AnyHashable: Any]?) -> Self? {
    guard let response = response else {
      return nil
    }
    let dict = (response as NSDictionary).stp_dictionaryByRemovingNulls() as NSDictionary
    let billingDetails = self.init()
    billingDetails.allResponseFields = response
    billingDetails.address = STPPaymentMethodAddress.decodedObject(
      fromAPIResponse: dict.stp_dictionary(forKey: "address"))
    billingDetails.email = dict.stp_string(forKey: "email")
    billingDetails.name = dict.stp_string(forKey: "name")
    billingDetails.phone = dict.stp_string(forKey: "phone")
    return billingDetails
  }
}
